### PR TITLE
Decouple substate tests from the global resolver.

### PR DESCRIPTION
**Pull request by patricksrobertson**
I've moved this into three different test modules:

* A default state that relies on the visit API.
* A nested route module that has shared setup for roughly half the tests
* An AutobootTestCase module for dealing with the one test case that
cannot be worked around.

I've removed the step testing function. This seemed to be more important
for debugging the authoring of the original sets of tests instead of
actually asserting code.

The instance of 'BRO' has been replaced with 'TURTLE' because I hate fun
and value inclusivity.

Refs #15058

### CODE REVIEW

File: packages/ember/tests/routing/substates_test.js [hunk 1]
```diff
@@ -1,1042 +1,880 @@
...
-  let broModel = {};
-  let broDeferred = RSVP.defer();
-
-  Router.map(function() {
-    this.route('bro');
-  });
-
-  App.ApplicationRoute = Route.extend({
-    setupController() {
-      step(2, 'ApplicationRoute#setup');
```
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Isn't some coverage lost by dropping these `step` assertions for what hooks fire in what order?

File: packages/ember/tests/routing/substates_test.js [hunk 2]
```diff
@@ -1,1042 +1,880 @@
...
+      }
+    }));
 
-  let deferred = RSVP.defer();
-  App.FooBarRoute = Route.extend({
-    model() {
-      return deferred.promise;
-    }
-  });
+    this.visit('/grandma/mom/sally');
```
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Please prefer the promise version of `visit`:

```js
return this.visit('/grandma/mom/sally').then(() => {
  let text = this.$('#app').text();
  /* ... */
});
```

File: packages/ember/tests/routing/substates_test.js [hunk 3]
```diff
@@ -1,1042 +1,880 @@
...
+      assert.equal(text, 'DUMMY', `dummy template has been rendered`);
+    });
+  }
 
-  App.IndexRoute = Route.extend({
-    model() {
-      return indexDeferred.promise;
-    }
-  });
+  ['@test Slow promises returned from ApplicationRoute#model enter ApplicationLoadingRoute if present - NEW'](assert) {
```
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Isn't this the same test?

* https://github.com/emberjs/ember.js/pull/15448/files#diff-a6a3e15b340e41ccc770054658e38485L733

Why `NEW` in the name?

File: packages/ember/tests/routing/substates_test.js [hunk 4]
```diff
@@ -1,1042 +1,880 @@
...
-  ok(appController.get('currentPath') !== 'loading', 'loading state not entered');
+    this.visit('/');
+    assert.ok(loadingRouteEntered, 'ApplicationLoadingRoute was entered');
 
-  run(indexDeferred, 'resolve', {});
-  equal(jQuery('#app', '#qunit-fixture').text(), 'INDEX');
-});
+    this.runTask(() => appDeferred.resolve());
 
-QUnit.test('Enter loading route if only LoadingRoute defined', function() {
```
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Is this test missing in the refactored file?

File: packages/ember/tests/routing/substates_test.js [hunk 5]
```diff
@@ -1,1042 +1,880 @@
...
-  }, function(err) { return err.msg === 'did it broke?'; });
+    return this.visit('/grandma/mom/sally').then(() => {
+      let appController = this.getController('application');
+      assert.equal(appController.get('currentPath'), 'grandma.mom.sally',
+        'Initial route fully loaded'
+      );
 
-  equal(jQuery('#app', '#qunit-fixture').text(), 'FOOBAR ERROR: did it broke?', 'foo.bar_error was entered (as opposed to something like foo/foo/bar_error)');
-});
+      this.visit('/grandma/smells');
```
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Should return the promise here and chain.

```js
  return this.visit('/grandma/smells');
}).then(() => {
  let appController = this.getController('application');
  assert.equal(appController.get('currentPath'), 'grandma.loading', /* ... */
```

File: packages/ember/tests/routing/substates_test.js [hunk 6]
```diff
@@ -1,1042 +1,880 @@
...
+        this.addTemplate('application_error', `
+          <p id="toplevel-error">TOPLEVEL ERROR: {{model.msg}}</p>
+        `);
+      })
+    }, (err) => err.msg === 'BAD NEWS BEARS', 'it went poorly');
+
+    let text = this.$('#toplevel-error').text();
+    assert.equal(text, 'TOPLEVEL ERROR: BAD NEWS BEARS',
+      'toplevel error rendered'
+    );
```
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Doesn't the original test attempt a `transitionTo` on the router after this?

* https://github.com/emberjs/ember.js/pull/15448/files#diff-a6a3e15b340e41ccc770054658e38485L1039

**Review by mixonic — CHANGES_REQUESTED** [review]
One or two things I would appreciate your thoughts on. Great start on a large migration, thank you.

I'd push your edits even further and swap `smells` for `cats` ;-)

File: packages/ember/tests/routing/substates_test.js [hunk 7]
```diff
@@ -1,1042 +1,880 @@
...
+  [`@test Loading actions bubble to root but don't enter substates above pivot `](assert) {
+    let sallyDeferred = RSVP.defer();
+    let smellsDeferred = RSVP.defer();
 
-QUnit.test('Slow promises returned from ApplicationRoute#model enter application_loading if template present', function() {
-  expect(3);
+    this.add('route:application', Route.extend({
+      actions: {
+        loading(transition, route) {
+          ok(true, 'loading action received on ApplicationRoute');
```
Comment (karthiick) [inline packages/ember/tests/routing/substates_test.js]: missed assert

File: packages/ember/tests/routing/substates_test.js (hunk 2 above)
Comment (patricksrobertson) [inline packages/ember/tests/routing/substates_test.js]: The problem is that the visit promise will not resolve until the model hook promise is resolved.

Is there another pattern to follow ie:
```js
let visit = this.visit('grandma/mom/sally');
// assertion before model is resolved
model.resolve();
visit.then(() => {
  //assertion after model hook is resolved
});
```

File: packages/ember/tests/routing/substates_test.js (hunk 5 above)
Comment (patricksrobertson) [inline packages/ember/tests/routing/substates_test.js]: same thing here as above ^, this relies on assertions before and after promises are resolved on `model()` hooks in routes.

File: packages/ember/tests/routing/substates_test.js (hunk 6 above)
Comment (patricksrobertson) [inline packages/ember/tests/routing/substates_test.js]: I can add that back in, my $0.02 is that you're testing behavior well tested in the basic routing test with that second assertion at the cost of pretty gnarly testing setup.

File: packages/ember/tests/routing/substates_test.js (hunk 1 above)
Comment (patricksrobertson) [inline packages/ember/tests/routing/substates_test.js]: Fair point!

File: packages/ember/tests/routing/substates_test.js (hunk 3 above)
Comment (patricksrobertson) [inline packages/ember/tests/routing/substates_test.js]: Shoot. The `- NEW` was when I was migrating tests over I could tell which of the two copies were failing. Will clean it out.

File: packages/ember/tests/routing/substates_test.js (hunk 4 above)
Comment (patricksrobertson) [inline packages/ember/tests/routing/substates_test.js]: https://github.com/emberjs/ember.js/pull/15448/files#diff-a6a3e15b340e41ccc770054658e38485R106

File: packages/ember/tests/routing/substates_test.js (hunk 2 above)
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Yeah that would works well. The thing I was reaching for was to chain the promises, and your example does that. Another alternative:

```js
let promise = this.visit('/grandma/mom/sally').then(() => {
  let text = this.$('#app').text();
});
model.resolve();
return promise;
```

File: packages/ember/tests/routing/substates_test.js (hunk 6 above)
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Yeah but it also is asserting that you can transition after an error substate. I'd prefer to add it back in.

File: packages/ember/tests/routing/substates_test.js [hunk 8]
```diff
@@ -1,1042 +1,951 @@
...
 
-  run(router, 'transitionTo', 'grandma.smells');
-  equal(appController.get('currentPath'), 'grandma.mom.sally', 'still in initial state because the only loading state is above the pivot route');
+        assert.equal(text, 'YAY', 'foo.bar.index fully loaded');
+      });
+      let text = this.$('#app').text();
 
-  run(smellsDeferred, 'resolve', {});
+      assert.equal(text, 'FOOBAR LOADING', `
+        foo.bar_loading was entered (as opposed to something like
```
Comment (mixonic) [inline packages/ember/tests/routing/substates_test.js]: Seems odd for an assertion message to start with a newline. Would prefer:

```js
  assert.equal(
    text, 'FOOBAR LOADING',
    `foo.bar_loading was entered (as opposed to something like foo/foo/bar_loading`
  );
```

**Review by mixonic — CHANGES_REQUESTED** [review]
Just some small cleanup and I think this is good to go.

**Comment by patricksrobertson** [comment]
@mixonic addressed!

**Review by mixonic — APPROVED** [review]

**Pull request merged**